app/assets/javascripts/components/AccountMenuReact.tsx

```diff
@@ -408,11 +412,74 @@ const AccountMenu = observer(({ application, appState, closeAccountMenu }: Props
   };
 
   const downloadDataArchive = () => {
-    console.log('downloadDataArchive');
+    application.getArchiveService().downloadBackup(isBackupEncrypted);
+  };
+
+
+  const readFile = async (file: File): Promise<any> => {
+    return new Promise((resolve) => {
+      const reader = new FileReader();
+      reader.onload = (e) => {
+        try {
+          const data = JSON.parse(e.target!.result as string);
+          resolve(data);
+        } catch (e) {
+          application.alertService.alert(STRING_INVALID_IMPORT_FILE);
+        }
+      };
+      reader.readAsText(file);
+    });
+  };
+
+  const performImport = async (data: BackupFile) => {
+    setIsImportDataLoading(true);
+
+    const result = await application.importData(data);
+
+    setIsImportDataLoading(false);
+
+    if (!result) {
+      return;
+    } else if ('error' in result) {
+      void alertDialog({
+        text: result.error,
+      });
+    } else if (result.errorCount) {
+      void alertDialog({
+        text: StringImportError(result.errorCount),
+      });
+    } else {
+      void alertDialog({
+        text: STRING_IMPORT_SUCCESS,
+      });
+    }
   };
 
-  const importFileSelected = () => {
-    console.log('importFileSelected');
+  const importFileSelected = async (event: TargetedEvent<HTMLInputElement, Event>) => {
+    const { files } = (event.target as HTMLInputElement);
+
+    if (!files) {
+      return;
+    }
+    const file = files[0];
+    const data = await readFile(file);
+    if (!data) {
+      return;
+    }
+    if (data.version || data.auth_params || data.keyParams) {
```

Comment:
could you turn this into something else? A cascade of if/else blocks with nested ifs can signify trouble in maintenance and debugging

Reply:
Done